### DISCUSSION TITLE
Where Implicit conversion operators from a base or abstract class makes sense

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#6892] [Language Ideas]
Dear all,
I would like to have support for operations on C# arrays such as:
```
x = alpha*y + beta*z + gamma*(w - z*y);
```
where x, y, z, w are C# `double[]` arrays  while alpha, beta, gamma are of type `double` and C# compiler replaces this expression with the hand coded loop,
```
for (int i = 0; i < y.Length; i++)
{
   x[i] = alpha*y[i] + beta*z[i] + gamma*(w[i] - z[i]*y[i]);
}
```

I know that the community that may need this is not that large, but many engineers and scientists in computational mathematics, physics, aeroastrophysics, and several other disciplines would benefit and would be attracted to C# since their codes are full of Linear Algebra operations. However, efficient Linear Algebra operations on native C# arrays and/or matrices, are not implemented by default. Other languages such as old and fast Fortran, Matlab, Julia allow this functionality by design, in C++ as well with the help of Template metaprogramming. Trying to implement something similar using operator overloading and smart inheritance I run up to the problem of the implicit conversion operator from a base class pointer restriction. I know that the standard does not allow it, but if you would like to extend a bit Csharp's arrays to perform Linear Algebra operations, at some point you have to assign an expression such as: 

```
double alpha, beta, gamma;
int n    = 1000;

Vector y = new Vector(n);
Vector z = new Vector(n);
Vector w = new Vector(n);

Vector x = alpha*y + beta*z + gamma*(w - z*y);
```
 At the moment what I am allowed to write is:
```
Vector x = new Vector( alpha*y + beta*z + gamma*(w - z*y) )
```
and this is enabled by a constructor taking an Expression `abstract class` as an argument
```
public class Vector
{
   public Vector(Expression expr) : base(e.Length)
   {
     expr.evaluate(this);
   }
}
```
I would also like to be able to declare the following constructor
```
public class Vector
{
   public static implicit operator Vector(Expression expr) 
   {
     return new Vector(expr);
   }
}
```
This would allow me to go ahead with the statement `Vector x = alpha*y + beta*z + gamma*(w - z*y);` There is a way at the moment to declare Vector as a generic class
```
public class Vector<T> where T : Expression
{
   public static implicit operator Vector<T>(T expr) 
   {
     return new Vector<T>(expr);
   }
}
```
but then all the declarations have to be written as:
```
Vector<Expression> y = new Vector<Expression>(n);
Vector<Expression> z = new Vector<Expression>(n);
Vector<Expression> w = new Vector<Expression>(n);

Vector<Expression> x = alpha*y + beta*z + gamma*(w - z*y);
```
which is really ugly. Clearly I would like to avoid this solution. Without luck I tried generic operators, something like:
```
public class Vector
{
   public static implicit operator Vector(T expr) where T : Expression
   {
     return new Vector(expr);
   }

   public static implicit operator<T> Vector(T expr) where T : Expression
   {
     return new Vector(expr);
   }
}
```
but the language does not allow me to do something like that. And all that due to the fact that Expression is a `pointer` to an `abstract class`. But you can clearly see why in this case it makes sense since it allows a syntax as close to mathematics on the paper as possible. 

Are there any smart ideas we could try out to bypass these restrictions? I am not so experienced with C# generics and all the bells and whistles of C# extension methods. 

Thank you very much for your time and effort.

**Comment by CyrusNajmabadi** [comment]
If you wrap your array in a struct, you can have all these facilities without any overhead. On phone, but can expand more in this if needed

**Reply by ghost** [reply]
Thank you for your answer, but it is not clear what you mean. Are you referring to the `double[]` array or my class `Vector` ? Furthermore, which  of the problems I mentioned is struct supposed to solve? Could you please be more specific?

**Reply by CyrusNajmabadi** [reply]
Def don't use a class. Just wrap the double[] in a struct. Then you can have operator overloading like you want.

**Reply by ghost** [reply]
Have you tried it? You know it is not working. You need to overload the operators for the struct and you are back to the same problem.

**Reply by CyrusNajmabadi** [reply]
Right. You get to overload the operators, so you can then write the code you want above.  (Maybe I missed how that would be a problem. Can you clarify?  Thanks!)

**Reply by ghost** [reply]
You end up with the same problem with classes. Nothing changes. Classes and structs are essentially the same.

**Reply by bernd5** [reply]
No, for structs you can add custom operators - for arrays not:

Practical [sample](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWACgsAGAAizwBYBuYsyvAOgGEB7AG04FMBjBFHYwAzvRIUqzADKwAjuMZSAcqjA9kUPmOK6iAWQCeANX4J2yADywEAPnIA3MxfIBecjB4B3cgG0AugDeeHDkOKEAzAC+4g4QyORgqJyCAA6cUDwAJm5h5ABUjs4JhTjixABmFjwQfAAWABRxCTDksInJaRnZAJSBxOSDTACcDTA9tEPEUXoiCGgC5EamAhaWACr2IEwRG/YDQ151Gjzk69sAkqpgwBp7xP1EQ+SpmnEIp+sB5FkQCBDiZ4HQZYCJLEzFBpffweNS3ZAiPrA54/P4QXIwOEaHRPIYzXEgsHLSFUXabWE3bFIgko37/DFYhHMdbsACCyGQEEMDQmyPxQJpoKYADY2mB0looAhyOxUho/i5iasrJsod9MZSEdSUUMsAB2DzeMaMxGAvHIoVUUVQcUZPhSmVyznmBJKl17BqyOZ7CnwxGPHUgg2eLzGzWmvnIi1gi4AUQ18vd5IA4jwEPG1ImLDy3PYGg1SXsenSID1mKn0wnndneTS41WFQlSeW0xn1NXkDnXPYSy3K5mOzyzYNoyLwSsk/ZZVmSg0zuQKrUXaE3WtyU5ldrA/qisqCgulxZh+R+VNBWCrePij7px2CnPVyr7Bvl/PF5vkQHA16ED6ECIMSNWtAwXapakaZoPDaVoXwsLcQMGSCQ2UDF93fF1jxA/9mFZLIsjGbxlGAwNTx1Hd/2PfEoiAA=):
```csharp
using System;
using System.Collections.Generic;
using System.Collections;
using System.Linq;
using System.Numerics;

MyVector<int> vector = new []{1, 2, 3};
var multiplied = 2 * vector * 2;
foreach(var n in multiplied){
    Console.WriteLine(n);   
}

struct MyVector<T> : IEnumerable<T> 
    where T: INumber<T>
{
    private T[] data;
    
    public MyVector(T[] numbers){
        data = numbers;
    }
    public MyVector(IEnumerable<T> numbers){
        data = numbers.ToArray();
    }
    
    public static implicit operator MyVector<T>(T[] numbers){
        return new(numbers);
    }
    public static implicit operator MyVector<T>(List<T> numbers){
        return new(numbers);
    }
    
    public IEnumerator<T> GetEnumerator() => ((IEnumerable<T>)data).GetEnumerator();
    IEnumerator IEnumerable.GetEnumerator() => data.GetEnumerator();
    
    public static MyVector<T> operator *(T factor, MyVector<T> vector){
        return vector * factor;
    }

    public static MyVector<T> operator *(MyVector<T> vector, T factor)
    {
        List<T> ts = new();
        foreach(var n in vector){
            var newN = n * factor;
            ts.Add(newN);
        }
        return ts;
    }
}
```

**Reply by ghost** [reply]
Unfortunately this implementation does not avoid the problem of temporary objects when you have to add lets say for example x1+x2+x3+x4. Every operator+ will have to add its 2 arguments and store the result somewhere. At the end it has to return a new MyVector<T>(data), where data[i]=x1[i]+x2[i]; You can try this out with array sizes around 10^8. You will see that the memory needed is beyond the memory for the 5 vectors x1,x2,x3,x4 and their sum.

**Reply by bernd5** [reply]
You are right with the implementation. But operator*() could return an `Expression` instead of a concrete `MyVector`.
The evaluation of such an expression can be handled in any way. E.g. for operator+ you could create a zero initialized sum array and just change the values of that array.

A sample (for multiply only):
```csharp

using System;
using System.Collections.Generic;
using System.Collections;
using System.Linq;
using System.Numerics;
using System.Drawing.Text;

MyVector<int> vector = new[] { 1, 2, 3 };
MyVector<int> multiplied = 2 * vector * 2;

foreach (var n in multiplied)
{
    Console.WriteLine(n);
}

public interface IMathVisitor<T, Result> where T : INumber<T>
{
    public Result Visit(MathExpression<T> expression);
    public Result VisitBinaryExpression(BinaryExpression<T> expression);
    public Result VisitVectorExpression(VectorExpression<T> expression);
    public Result VisitScalarExpression(ScalarExpression<T> expression);
}

public class MathEvalVisitor<T> : 
    IMathVisitor<T, SimpleExpression<T>> where T: INumber<T>
{
    public SimpleExpression<T> Visit(MathExpression<T> expression)
    {
        return expression.Accept(this);
    }

    public SimpleExpression<T> VisitBinaryExpression(BinaryExpression<T> expression)
    {
        //here we can combine scalars and vectors
        var leftResult = Visit(expression.Left);
        var rightResult = Visit(expression.Right);

        switch((leftResult, rightResult))
        {
            case (ScalarExpression<T> a, ScalarExpression<T> b):
            {
                    return new ScalarExpression<T>
                    {
                        Value = a.Value * b.Value
                    };
            }
            case (VectorExpression<T> a, ScalarExpression<T> b):
            {
                return DoMultiply(a, b);
            }
            case (ScalarExpression<T> b, VectorExpression<T> a):
            {
                return DoMultiply(a, b);
            }
            default:
                throw new NotImplementedException();
        }

        SimpleExpression<T> DoMultiply(VectorExpression<T> vector, ScalarExpression<T> factor)
        {
            List<T> ts = new();
            foreach (var n in vector.Vector)
            {
                T newN = n * factor.Value;
                ts.Add(newN);
            }
            MyVector<T> v = ts;
            VectorExpression<T> result = v;
            return result;
        }
    }


    public SimpleExpression<T> VisitVectorExpression(VectorExpression<T> expression)
    {
        return expression;
    }
    public SimpleExpression<T> VisitScalarExpression(ScalarExpression<T> expression)
    {
        return expression;
    }
}


public abstract record MathExpression<T> where T : INumber<T>
{
    public abstract Result Accept<Result>(IMathVisitor<T, Result> visitor);

    public static implicit operator MathExpression<T>(MyVector<T> values)
    {
        return new VectorExpression<T>()
        {
            Vector = values
        };
    }
    public static implicit operator MathExpression<T>(T value)
    {
        return new ScalarExpression<T>()
        {
            Value = value
        };
    }

    public static MathExpression<T> operator *(MathExpression<T> left, MathExpression<T> right)
    {
        var expr = new BinaryExpression<T>
        {
            Left = left,
            Right = right,
            Kind = BinaryExpressionKind.Multiply
        };
        return expr;
    }

    public static implicit operator MyVector<T>(MathExpression<T> expr)
    {
        var eval = new MathEvalVisitor<T>();
        var r = eval.Visit(expr);
        switch (r)
        {
            case VectorExpression<T> v:
                return v.Vector;
            default:
                throw new Exception("invalid math");
        }
    }
}

public abstract record SimpleExpression<T> : MathExpression<T>
    where T : INumber<T>
{

}

public record VectorExpression<T> : SimpleExpression<T>
    where T : INumber<T>
{
    public MyVector<T> Vector { get; init; }

    public override Result Accept<Result>(IMathVisitor<T, Result> visitor)
    {
        return visitor.VisitVectorExpression(this);
    }

    public static implicit operator VectorExpression<T>(MyVector<T> values)
    {
        return new VectorExpression<T>()
        {
            Vector = values
        };
    }
}

public record ScalarExpression<T> : SimpleExpression<T>
    where T : INumber<T>
{
    public T Value { get; init; }

    public override Result Accept<Result>(IMathVisitor<T, Result> visitor)
    {
        return visitor.VisitScalarExpression(this);
    }
}


public enum BinaryExpressionKind
{
    Multiply,
    Add,
    Subtract,
    Divide
}

public record BinaryExpression<T> : MathExpression<T>
    where T : INumber<T>
{
    public BinaryExpressionKind Kind { get; init; }

    public MathExpression<T> Left { get; init; }
    public MathExpression<T> Right { get; init; }

    public override Result Accept<Result>(IMathVisitor<T, Result> visitor)
    {
        return visitor.VisitBinaryExpression(this);
    }
}


public struct MyVector<T> : IEnumerable<T>
    where T : INumber<T>
{
    private T[] data;

    public MyVector(T[] numbers)
    {
        data = numbers;
    }
    public MyVector(IEnumerable<T> numbers)
    {
        data = numbers.ToArray();
    }

    public static implicit operator MyVector<T>(T[] numbers)
    {
        return new(numbers);
    }
    public static implicit operator MyVector<T>(List<T> numbers)
    {
        return new(numbers);
    }

    public IEnumerator<T> GetEnumerator() => ((IEnumerable<T>)data).GetEnumerator();
    IEnumerator IEnumerable.GetEnumerator() => data.GetEnumerator();

    public static MathExpression<T> operator *(T factor, MyVector<T> vector)
    {
        MathExpression<T> left = vector;
        MathExpression<T> right = factor;

        return left * right;
    }

    public static MathExpression<T> operator *(MyVector<T> vector, T factor)
    {
        MathExpression<T> left = vector;
        MathExpression<T> right = factor;

        return left * right;
    }
}
```

Before `MathEvalVisitor` the tree could be rewritten to avoid temporaries...
The code for Accept, VisitXXX... I would generate with a source generator.
And it misses the check for binary operator kind...
But I hope you get the point.

**Reply by ghost** [reply]
You have some classes and some records. What is the reason of using records?

> Before MathEvalVisitor the tree could be rewritten to avoid temporaries...
How can you rewrite the tree to avoid temporaries?

I tried to create expressions in a simpler way without intermediate temporaries. Both `PlusExpr `or `MulExpr `inherit however from an Interface which defines a `Value(i)` method. Then in order to test the performance I explicitly wrote the expression myself as:
```
// x = x1 + x2 + x3*x4
MyVector x = Add(Add(x1,x2), Mul(x3,x4));
```
Although temporary objects are avoided this way, the runtime performance is about 10 times slower than the  hand-coded loop due to the indirections imposed by inheritance from the Interface (which enforces Value(i)).

> The code for Accept, VisitXXX... I would generate with a source generator.

How to employ a source generator on that? I have no idea on how to call the compiler to do such stuff. But wouldn't this have a runtime overhead? I mean if you call the compiler to compile the code for each expression before evaluating the expression, then the compile time would be added to the runtime performance and for relative small arrays (n < 10,000,000) the compilation time would be 100 times or more slower than the time needed to execute the `for `loop.

**Reply by bernd5** [reply]
I use records for the expressions to have both immutablity and easy rewriting. E.g. if you can simpify the left part of a binary expression (i would prefer that with a kind enum instead of MulExpr...) you can write:
```csharp
var adjustedBinaryExpr = oldBinary with { Left = optimizedLeft };
```
You may have a look at `roslyn` - the csharp reference implementation.
They use there immutable objects which have "withXXX" methods to create a new object with the adjusted properties.

Not sure how your AddExpr looks. But the idea is to create a tree of expressions. Then you can modify that tree (e.g. combine scalar mulitplication).
After that you can evaluate the expression tree. During evaluation you can analyze the full expression tree - e.g. that you have multiple vectors.

Of course the expression tree has some overhead, but it is nothing compared to over 10,000,000 math operations (which is huge in my eyes).

The compile time overhead is only relevant during compilation - no runtime overhead (maybe a bit more work for the Jitter - but just once for every function).

In the sample above I used the "visitor pattern" - which is very great for tree traversal and often used by compilers. To make that work we need to add an `Accept` function for every visitable node and a corresponding interface method. In addition we could write a "rewriter" which would have default implementations for every node, e.g.:

```csharp
public MathExpression<T> VisitBinaryExpression(BinaryExpression expression)
{
   var adjustedLeft = Visit(expression.Left);
   var adjustedRight = Visit(expression.Right);
   if (ReferenceEquals(expression.Left, adjustedLeft) && ReferenceEquals(expression.Right, adjustedRight))
   {
      return expression;
   }
  return expression with { Left = adjustedLeft, Right = adjustedRight };
}
```

Such code could be generated by source generator because it follows a pattern...
